### PR TITLE
Recorded Future Identity Module Integration with XSOAR

### PR DESCRIPTION
**Pull request by rderkachrf**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Description
New integration with Recorded Future Identity Module

## Screenshots
image
image

## Minimum version of Cortex XSOAR
- [x] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: Packs/IdentityRecordedFuture/Integrations/IdentityRecordedFuture/example_commands.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+!ip ip="8.8.8.8"
```
Comment (dorschw) [inline Packs/IdentityRecordedFuture/Integrations/IdentityRecordedFuture/example_commands.txt:1]: seems like template leftovers, please either supply relevant commands or delete the file

**Comment by dorschw** [comment]
Hi, 
Regarding the `BA102` you're getting in the validations, please move any non-XSOAR file (test data mostly) to a folder called `test_data`, to get the CI to ignore them.

File: Packs/IdentityRecordedFuture/.secrets-ignore [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+12000$IWIs9x2tM7U3afasdasd3d23nRfo0sFRjpkMlim2GA2+p/2Y7RQLpODP4S0=
+bbebf4e24e6631570cd8f60e1b0f77c
+7bf5626ca4944595aa89bf5fdsdfsdf9c9a1e9b014356d9b10ef31c9e9aadc7835e4
+a6b9607b296pfk941dd390d9e690ade8e
+6204f1659eb8dasdas871be7f4b3ee560207bda2c5
+a2abf21c7669d58137kffsa83381d55cc
+__fake1@fake.com
+Cortex_XSOAR/2.0
+Cortex_XSOAR_unittest_0.1
+realname@fake.com
```
Comment (dorschw) [inline Packs/IdentityRecordedFuture/.secrets-ignore:12]:
```suggestion
realname@fake.com
notreal@notfake.com
https://id.notfake.com
fake1@fake1.com
fake12@fake1.com
fake13@fake1.com
fake14@fake1.com
```

add these to solve CI validation failures

**Review by dorschw — CHANGES_REQUESTED** [review]

**Comment by edik24** [comment]
Hi, there is an error with lint tests.
you need to pull from the master again and then commit to resolve the current error

**Comment by rderkachrf** [comment]
@edik24 
I did it wrong.
I rebase my branch to master, and then force pushed.
That is a huge mistake.

**Comment by dorschw** [comment]
> @edik24 I did it wrong. I rebase my branch to master, and then force pushed. Not the best idea. Used git reflog and returned everything to its original state. Is there guidance how to pull from the master again and then commit to resolve

you can use [this script](https://github.com/demisto/content/blob/master/Utils/git_pull_master_into_fork.sh) for easily merging content/master into your own fork.

File: Packs/IdentityRecordedFuture/Integrations/IdentityRecordedFuture/IdentityRecordedFuture.py [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+                for idx, credential in enumerate(credentials):
+                    exposed_secret = credential["exposed_secret"]
+                    password_number_text = f"__Password {idx + 1}__"
+                    passwords_section.append(f"{password_number_text}:")
+                    if exposed_secret.get("details", {}).get("rank"):
+                        passwords_section.append(
+                            f"__Rank__: {exposed_secret['details']['rank']}"
+                        )
+                    if exposed_secret[
+                        "effectively_clear"
```
Comment (dorschw) [inline Packs/IdentityRecordedFuture/Integrations/IdentityRecordedFuture/IdentityRecordedFuture.py:251]: is `effectively_clear` always a key of `exposed_secret`? 

if not, it's recommended to use `exposed_secret.get("effectively_clear",<True or False>)` instead.

File: Packs/IdentityRecordedFuture/Integrations/IdentityRecordedFuture/IdentityRecordedFuture.py [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+                    ] and exposed_secret.get("details", {}).get(
+                        "clear_text_hint"
+                    ):
+                        passwords_section.append(
+                            f"{exposed_secret['details']['clear_text_hint']} (__{exposed_secret['type']}__)"
+                        )
+                    if exposed_secret["hashes"]:
```
Comment (dorschw) [inline Packs/IdentityRecordedFuture/Integrations/IdentityRecordedFuture/IdentityRecordedFuture.py:258]: see `effectively_clear` comment.

File: Packs/IdentityRecordedFuture/Integrations/IdentityRecordedFuture/IdentityRecordedFuture.py [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+                        ).strftime("%b %Y")
+                        passwords_section.append(
+                            f"Exfiltration date: {exfiltration_date}"
+                        )
+                    for dump in credential["dumps"]:
+                        dump_downloaded = dump.get("downloaded", "")
+                        if dump_downloaded:
+                            dump_downloaded = datetime.strptime(
+                                dump_downloaded, ISO_DATE_FORMAT
+                            ).strftime("%b %Y")
```
Comment (dorschw) [inline Packs/IdentityRecordedFuture/Integrations/IdentityRecordedFuture/IdentityRecordedFuture.py:285]: please use `dateparser.parse` instead of `strptime`

File: Packs/IdentityRecordedFuture/Integrations/IdentityRecordedFuture/IdentityRecordedFuture.py [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+                                dumps_section.append(
+                                    f"__Dump type:__ {dump['type']}"
+                                )
+                            dumps_section.append(
+                                f"{dump.get('description', '')}"
+                            )
+                        for breach in dump.get("breaches", []):
+                            breach_date = breach.get("breached")
+                            if breach_date:
+                                breach_date = datetime.strptime(
```
Comment (dorschw) [inline Packs/IdentityRecordedFuture/Integrations/IdentityRecordedFuture/IdentityRecordedFuture.py:304]: please use `dateparser.parse` instead of `strptime`

File: Packs/IdentityRecordedFuture/Integrations/IdentityRecordedFuture/IdentityRecordedFuture.py [hunk 7]
```diff
@@ -0,0 +1,437 @@
...
+        if command == "test-module":
+            try:
+                client.whoami()
+                return_results("ok")
+            except Exception as err:
+                message = str(err)
+                try:
+                    error = json.loads(str(err).split("\n")[1])
+                    if "fail" in error.get("result", {}).get("status", ""):
+                        message = error.get("result", {})["message"]
```
Comment (dorschw) [inline Packs/IdentityRecordedFuture/Integrations/IdentityRecordedFuture/IdentityRecordedFuture.py:401]:
```suggestion
                        message = error.get("result", {}).get("message",<some default message>)
```

to avoid KeyError on accessing `{}["message"]`


on a second thought, catching it on the next line like you did, is also an option, both are good.

File: Packs/IdentityRecordedFuture/Integrations/IdentityRecordedFuture/IdentityRecordedFuture.yml [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+      defaultValue: "All time"
+      description: Time period for last downloaded time.
+      predefined:
+      - "Last 24 Hours"
+      - "Last 7 Days"
+      - "Last Month"
+      - "Last 3 Months"
+      - "Last 6 Months"
+      - "Last Year"
+      - "All time"
```
Comment (dorschw) [inline Packs/IdentityRecordedFuture/Integrations/IdentityRecordedFuture/IdentityRecordedFuture.yml:78]: Does the API only support these values? or can any period be queried?

File: Packs/IdentityRecordedFuture/Integrations/IdentityRecordedFuture/IdentityRecordedFuture.yml [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
```
Comment (dorschw) [inline Packs/IdentityRecordedFuture/Integrations/IdentityRecordedFuture/IdentityRecordedFuture.yml:112]: Does the API only support these values? or can any period be queried?

**Review by dorschw — CHANGES_REQUESTED** [review]
Good job, let's schedule a demo over DFIR.

File: Packs/IdentityRecordedFuture/Integrations/IdentityRecordedFuture/IdentityRecordedFuture.py (hunk 3 above)
Comment (rderkachrf) [reply Packs/IdentityRecordedFuture/Integrations/IdentityRecordedFuture/IdentityRecordedFuture.py:251]: It is. But just to protect the code I will use `get`

File: Packs/IdentityRecordedFuture/Integrations/IdentityRecordedFuture/IdentityRecordedFuture.py (hunk 6 above)
Comment (rderkachrf) [reply Packs/IdentityRecordedFuture/Integrations/IdentityRecordedFuture/IdentityRecordedFuture.py:304]: Cool, didn't know about that lib

File: Packs/IdentityRecordedFuture/Integrations/IdentityRecordedFuture/IdentityRecordedFuture.yml (hunk 8 above)
Comment (rderkachrf) [reply Packs/IdentityRecordedFuture/Integrations/IdentityRecordedFuture/IdentityRecordedFuture.yml:78]: API support any date but to make the User Interface easier to use we decided to have periods predefined.

Comment (dorschw) [reply Packs/IdentityRecordedFuture/Integrations/IdentityRecordedFuture/IdentityRecordedFuture.yml:78]: Our recommendation is to allow a flexible querying when the server supports it: use a free text (type 0) argument field, and call dateparser.parse on it. This subtracts the input from the current time and returns a `datetime.datetime` object:

```python
>>> parse('3 years')
datetime.datetime(2019, 1, 26, 14, 33, 37, 196384)
```

"all time" is the only thing that won't be automatically parsed by dateparser - to go around it, it's possible that you consider an empty input as all times, and check for it before calling `dateparser.parse`. Just make sure to tell the users (in either the `additional info` or `display` fields of the argument, in the yaml) to make it easier for them to use it.

File: Tests/conf.json [hunk 10]
```diff
@@ -5460,4 +5460,4 @@
             }
         }
     }
-}
\ No newline at end of file
+}
```
Comment (dorschw) [inline Tests/conf.json:5463]: please undo this change (added a newline in the end of the file)

Comment (rderkachrf) [reply Tests/conf.json]: done.

**Review by DeanArbel — APPROVED** [review]
@rderkachrf Thank you very much for this contribution. I will now start doing our internal testing, and if everything is ok, the changes should be up by tomorrow.
